### ISSUE TITLE
Generate a Code Signing Certificate

### ISSUE THREAD

**Issue by CMCDragonkai** [smallstep/certificates#147]
### What would you like to be added

I would like to be able to generate a code signing x509 certificate. This adds a new extended key usage constraint called Code Signing.

With openssl it's like this:

```
  openssl req -newkey rsa -keyout codesign.key -out codesign.req
  openssl ca -config ca.cnf -extensions codesigning -in codesign.req -out codesign.crt
  openssl cms -sign -binary -noattr -in vmlinuz \
              -signer codesign.crt -inkey codesign.key -certfile ca.crt \
              -outform DER -out vmlinuz.sig
```

### Why this is needed

This allows x509 certs to be used to sign digital artifacts.

I checked the `step ca certificate` command and couldn't find an option to enable this.

**Comment by dopey** [comment]
Hey @CMCDragonkai, thanks for opening the issue. Unfortunately there is currently no option to enable to enable the code-signing extension. We have an existing issue open (https://github.com/smallstep/cli/issues/110) on the `cli` repo that would enable a much wider array of features. It's a common request, and we're planning on working on this in short order in the new year.

In the mean time ... If you're able to build `step` locally, then this should be a very easy change to make. 
Just add `x509.ExtKeyUsageCodeSigning` to the following code block https://github.com/smallstep/cli/blob/master/crypto/x509util/leafProfile.go#L76-L79. This will make it so that all new `leaf` certificates have the code-signing extension enabled.  Since this change is in `cli` you would need to update your build of `certificates` to use your branch of `cli` (instead of the smallstep` one). Let me know if you have questions about this and I'll be happy to help.

tl;dr these features will be coming in a release soon; but if you'd like them now, then you can pull the code, make a one line change, build, and be good to go.

**Comment by CMCDragonkai** [comment]
Cool.

Do you have an ETA for the next release then?

**Comment by CMCDragonkai** [comment]
Another issue is that generating the code signing cert is only 1 thing. Another is signing things with the code signing cert. This requires using the `cms` command in openssl but it's pretty clunky. Is it also in step's plans to add these commands in as well?

**Comment by mmalone** [comment]
Broadly, code signing is something we're interested in supporting at some point. So, yes, absolutely, we'd like to make it easy for users to actually sign code.

One thing slowing this work (apart from just having more things to do than time to do it :) is a general lack of experience / exposure to code signing use cases in practice. If/when we lean into code signing, we'd like to provide tools that work at the appropriate level of abstraction to safely and easily satisfy common use cases. We don't want to implement another clunky tool.

I'm guessing the most basic requirement would be:

 * A command to sign a piece of code with a private key and construct a bundle containing the code, the signature, and the (code signing) certificate associated with the private key.
 * A command that takes this bundle and a root certificate as input, extracts and verifies the signature and certificate, and reports whether the signature is valid (or maybe just extracts the code).

This would satisfy a requirement like "put a technical mechanism in place that allows servers to determine whether code has been through the production deploy pipeline". Is this your use case?

Help identifying any additional common use cases would be very much appreciated (e.g., signing code for an app store, maybe?). Same with identifying relevant standards (e.g., is there a standard way to package the `(code, signature, certificate)` tuple described above) and integrations (e.g., are there things we can/should to do make it easier for this to work with things like Puppet/Chef/Ansible/Terraform/Kubernetes).

Anyone reading this who wants this stuff, please reply here if you have any thoughts!

**Comment by mmalone** [comment]
Thanks. That's helpful.

I wonder if you could just use the JWS (JSON Web Signature) functionality `step` already has for this? I think it _can_ do what you need. But I'm guessing it's not quite what you want. Still, it'll help clarify requirements.

This is a bit of a stream-of-consciousness brain dump, and it's using some fairly advanced features, so lemme know if anything doesn't make sense and I can help clarify.

Here's how it could work:

## Signing some code using JWS

First, create a JWK (JSON Web Key) (a key pair):

```
$ step crypto jwk create k.pub k.priv
Please enter the password to encrypt the private JWK:
Your public key has been saved in k.pub.
Your private key has been saved in k.priv.
```

Then hash the code you're signing and sign the hash, generating a JWS (I'm just hashing the `step` binary in this demo):

```
$ step crypto hash digest step | step crypto jws sign --key k.priv
Please enter the password to decrypt k.priv:
eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw
```

If we inspect this artifact, we'll see that it has a payload containing our checksum:

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw | step crypto jws inspect --insecure
9364cd2b69296972b0eeabe24f9516336676da45689507655c2f1db9ea69d1b4  step
```

And some metadata containing key information and a signature over the checksum file:

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw | step crypto jws inspect --insecure --json
{
  "header": {
    "alg": "ES256",
    "kid": "7tGCpp8lZ2Yv4hz0Y0LH4Fmq3AW4W-3zXsjRoopE5PQ"
  },
  "payload": "OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo",
  "signature": "V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw"
}
```

## Verifying the signature

If you distribute this JWS alongside the code, and the recipient has your public key, they can verify the signature:

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw | step crypto jws verify --key k.pub
9364cd2b69296972b0eeabe24f9516336676da45689507655c2f1db9ea69d1b4  step
```

The output here is the checksum structure generated by `step crypto hash digest`. As a final step, we can pipe this into `step crypto hash compare` to check the signature:

```
echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw \
  | step crypto jws verify --key k.pub \
  | awk '{print $1}' \
  | xargs -Istep-hash step crypto hash compare step-hash step
ok
```

Both `step crypto jws verify` and `step crypto hash compare` return zero on success:

```
$ echo $?
0
```

and return non-zero if there's a hash mismatch or signature verification failure, respectively. So you can check `$?` to see if this pipeline completed successfully (indicating that the code is properly signed):

Example of bad JWS signature:

```
$ (
>   set -o pipefail
>   echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.BADSIG | step crypto jws verify --key k.pub | awk '{print $1}' | xargs -Istep-hash step crypto hash compare step-hash step
> )
validation failed: invalid signature
$ echo $?
1
```

(Note that the parens contain the effects of the `pipefail` setting, which ensures that the entire pipeline fails with a non-zero exit code when one of the subcommands fails).

Example of bad checksum:

```
$ (
>   set -o pipefail
>   echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw | step crypto jws verify --key k.pub | awk '{print $1}' | step crypto hash compare abc123 step
> )
fail
$ echo $?
1
```

## Key distribution

The problem with JWKs is that the recipient of your code needs to know your public key in order to verify the signature. Maybe you already have a way to distribute keys. If not, here are two options.

### Using JWK Sets

First, JWKs can be arranged into sets. Pointing a verification routine at a JWK set, rather than a single JWK, makes it easier to rotate keys. OAuth OIDC uses JWK sets to verify identity tokens, for example (Google serves their JWK public keys for verifying identity tokens [here](https://www.googleapis.com/oauth2/v3/certs)).

You can use `step` to create your own JWK set:

```
$ cat k.pub | step crypto jwk keyset add keys.json
```

Then get a (normal TLS) cert and serve it using the (hidden) `step` fileserver (or whatever server you want to use):

```
$ step ca certificate localhost l.crt l.key
✔ Provisioner: mike@example.com (JWK) [kid: 6PTCmbhoB2e0ffUz0mQBfk3ntj3STcFxzDZyWr8yHaY]
✔ Please enter the password to decrypt the provisioner key:
✔ CA: https://127.0.0.1
✔ Certificate: l.crt
✔ Private Key: l.key
$ step fileserver --cert l.crt --key l.key .
Serving HTTPS at [::]:64683 ...
```

To verify our JWS signature, we can fetch the currently valid JWKs from this server using HTTPS:

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6Ijd0R0NwcDhsWjJZdjRoejBZMExINEZtcTNBVzRXLTN6WHNqUm9vcEU1UFEifQ.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.V94GlYeQQ72lLhzYUYIaXikZ9WhB7kEQGjkiUEpqBKNwZovp11sI2wF9N92SY1S5xwUHim-eJleAd0cIrBAsHw \
>   | step crypto jws verify --jwks <(curl -s --cacert $(step path)/certs/root_ca.crt https://localhost:64683/keys.json) \
>   | awk '{print $1}' \
>   | xargs -Istep-hash step crypto hash compare step-hash step
ok
```

### Using `x5c` (certificates)

Another option for key distribution is to actually _use certificates_. A JWS can have an `x5c` parameter that contains a certificate or certificate chain.

Create a certificate & key for signing your JWS:

```
step ca certificate code-signer c.crt c.key
```

Generate your JWS, including the cert bundle from `c.crt` in the `x5c` parameter and using `c.key` to sign:

```
$ step crypto hash digest step | step crypto jws sign --x5c-cert c.crt --key c.key
eyJhbGciOiJFUzI1NiIsImtpZCI6ImE3RW5ZbEI1QXp1S184ckNOZUlPZVJ2blRpU2VwM3Z6TXkwVDc1VzFBelEiLCJ4NWMiOlsiTUlJQ0ZUQ0NBYnVnQXdJQkFnSVJBS2Q2WUFpU0s1U2pZTEN2dnZxQnhvc3dDZ1lJS29aSXpqMEVBd0l3SHpFZE1Cc0dBMVVFQXhNVVZHVnpkQ0JKYm5SbGNtMWxaR2xoZEdVZ1EwRXdIaGNOTWpBd01UQTVNakF6T0RJMVdoY05NakF3TVRFd01qQXpPREkxV2pBV01SUXdFZ1lEVlFRREV3dGpiMlJsTFhOcFoyNWxjakJaTUJNR0J5cUdTTTQ5QWdFR0NDcUdTTTQ5QXdFSEEwSUFCRlN5SDFyekNVeHFCbXY0aVd2MTRmRTdQT0ROc2hsb0ljZ0l3Y2dpTk90MGhqcCtGNWl0QkJWaTlPM1Y1SXc5MFJWWG9VazhERmdjOVJPUjQybEpHN1NqZ2VBd2dkMHdEZ1lEVlIwUEFRSC9CQVFEQWdXZ01CMEdBMVVkSlFRV01CUUdDQ3NHQVFVRkJ3TUJCZ2dyQmdFRkJRY0RBakFkQmdOVkhRNEVGZ1FVQ0xrVFNoN2wyUE9HU2xjRzdrcVlvc1czMnhvd0h3WURWUjBqQkJnd0ZvQVVjaXVWbHNDSzJGYzV6K1hTcC91UCtKVjMzdlV3RmdZRFZSMFJCQTh3RFlJTFkyOWtaUzF6YVdkdVpYSXdWQVlNS3dZQkJBR0NwR1RHS0VBQkJFUXdRZ0lCQVFRUWJXbHJaVUJsZUdGdGNHeGxMbU52YlFRck5sQlVRMjFpYUc5Q01tVXdabVpWZWpCdFVVSm1hek51ZEdvelUxUmpSbmg2UkZwNVYzSTRlVWhoV1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThHYUdNcDJBbFJTcFB3MEQ0ZGVpaGw4RjJlczlHMHZIMVFHQnl6bG5mUUFDSUhYcXoyL1c2TTJkd1dUdC9TUzc3SElueEU3WklWdmhLa3hDcW1kRUtIdS8iLCJNSUlCbWpDQ0FVQ2dBd0lCQWdJUWZTMVlFNVRvMUFxQkNFYzN5U2RYTURBS0JnZ3Foa2pPUFFRREFqQVhNUlV3RXdZRFZRUURFd3hVWlhOMElGSnZiM1FnUTBFd0hoY05NakF3TVRBNU1UZzFPVEUxV2hjTk16QXdNVEEyTVRnMU9URTFXakFmTVIwd0d3WURWUVFERXhSVVpYTjBJRWx1ZEdWeWJXVmthV0YwWlNCRFFUQlpNQk1HQnlxR1NNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQUJKRDdta25GSTJpQkJRT3RzdmtDdGs1Qm4xUitOWkdEaG11WENXUWs5RHJyYjRzMXJjZ3BPZy9OSHV2VlpUcGZMeGpzWEdtVk9RYWJMWnFHWG9RbDZwYWpaakJrTUE0R0ExVWREd0VCL3dRRUF3SUJCakFTQmdOVkhSTUJBZjhFQ0RBR0FRSC9BZ0VBTUIwR0ExVWREZ1FXQkJSeUs1V1d3SXJZVnpuUDVkS24rNC80bFhmZTlUQWZCZ05WSFNNRUdEQVdnQlNZdlozbVRHZDF5T2Zrc09QUzV4U2VBdEtkZ1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThJU25reVo4QXVQT3Q1NTNVZnpMSDBBS2ZuYytqTDhxN0ozdUVYbkRCUmNDSUdWVWN2OE1uV0VXS0tPNUpaQitqVDlpV3d5KzJPZm5tbHRrNU45VHFRQnUiXX0.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.6IwpHrrlelrZainCS5fKPehbTnA53lLBQwVvKKFqiSd0HfIQFmcelrK6uURjTVR1PWruxhIShFduAQaObupfSQ
```

The recipient of this JWS can extract the certificate chain from the `x5c` parameter (this is a bit gnarly, but it works):

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6ImE3RW5ZbEI1QXp1S184ckNOZUlPZVJ2blRpU2VwM3Z6TXkwVDc1VzFBelEiLCJ4NWMiOlsiTUlJQ0ZUQ0NBYnVnQXdJQkFnSVJBS2Q2WUFpU0s1U2pZTEN2dnZxQnhvc3dDZ1lJS29aSXpqMEVBd0l3SHpFZE1Cc0dBMVVFQXhNVVZHVnpkQ0JKYm5SbGNtMWxaR2xoZEdVZ1EwRXdIaGNOTWpBd01UQTVNakF6T0RJMVdoY05NakF3TVRFd01qQXpPREkxV2pBV01SUXdFZ1lEVlFRREV3dGpiMlJsTFhOcFoyNWxjakJaTUJNR0J5cUdTTTQ5QWdFR0NDcUdTTTQ5QXdFSEEwSUFCRlN5SDFyekNVeHFCbXY0aVd2MTRmRTdQT0ROc2hsb0ljZ0l3Y2dpTk90MGhqcCtGNWl0QkJWaTlPM1Y1SXc5MFJWWG9VazhERmdjOVJPUjQybEpHN1NqZ2VBd2dkMHdEZ1lEVlIwUEFRSC9CQVFEQWdXZ01CMEdBMVVkSlFRV01CUUdDQ3NHQVFVRkJ3TUJCZ2dyQmdFRkJRY0RBakFkQmdOVkhRNEVGZ1FVQ0xrVFNoN2wyUE9HU2xjRzdrcVlvc1czMnhvd0h3WURWUjBqQkJnd0ZvQVVjaXVWbHNDSzJGYzV6K1hTcC91UCtKVjMzdlV3RmdZRFZSMFJCQTh3RFlJTFkyOWtaUzF6YVdkdVpYSXdWQVlNS3dZQkJBR0NwR1RHS0VBQkJFUXdRZ0lCQVFRUWJXbHJaVUJsZUdGdGNHeGxMbU52YlFRck5sQlVRMjFpYUc5Q01tVXdabVpWZWpCdFVVSm1hek51ZEdvelUxUmpSbmg2UkZwNVYzSTRlVWhoV1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThHYUdNcDJBbFJTcFB3MEQ0ZGVpaGw4RjJlczlHMHZIMVFHQnl6bG5mUUFDSUhYcXoyL1c2TTJkd1dUdC9TUzc3SElueEU3WklWdmhLa3hDcW1kRUtIdS8iLCJNSUlCbWpDQ0FVQ2dBd0lCQWdJUWZTMVlFNVRvMUFxQkNFYzN5U2RYTURBS0JnZ3Foa2pPUFFRREFqQVhNUlV3RXdZRFZRUURFd3hVWlhOMElGSnZiM1FnUTBFd0hoY05NakF3TVRBNU1UZzFPVEUxV2hjTk16QXdNVEEyTVRnMU9URTFXakFmTVIwd0d3WURWUVFERXhSVVpYTjBJRWx1ZEdWeWJXVmthV0YwWlNCRFFUQlpNQk1HQnlxR1NNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQUJKRDdta25GSTJpQkJRT3RzdmtDdGs1Qm4xUitOWkdEaG11WENXUWs5RHJyYjRzMXJjZ3BPZy9OSHV2VlpUcGZMeGpzWEdtVk9RYWJMWnFHWG9RbDZwYWpaakJrTUE0R0ExVWREd0VCL3dRRUF3SUJCakFTQmdOVkhSTUJBZjhFQ0RBR0FRSC9BZ0VBTUIwR0ExVWREZ1FXQkJSeUs1V1d3SXJZVnpuUDVkS24rNC80bFhmZTlUQWZCZ05WSFNNRUdEQVdnQlNZdlozbVRHZDF5T2Zrc09QUzV4U2VBdEtkZ1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThJU25reVo4QXVQT3Q1NTNVZnpMSDBBS2ZuYytqTDhxN0ozdUVYbkRCUmNDSUdWVWN2OE1uV0VXS0tPNUpaQitqVDlpV3d5KzJPZm5tbHRrNU45VHFRQnUiXX0.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.6IwpHrrlelrZainCS5fKPehbTnA53lLBQwVvKKFqiSd0HfIQFmcelrK6uURjTVR1PWruxhIShFduAQaObupfSQ \
  | step crypto jws inspect --json --insecure \
  | jq -r .header.x5c[] \
  | awk '$0="-----BEGIN CERTIFICATE-----\n"$0"\n-----END CERTIFICATE-----"' \
  | fold -w 64 > signer.crt
```

Verify the chain using your root certificate:

```
$ step certificate verify signer.crt --host code-signer --roots $(step path)/certs/root_ca.crt
```

(Again, this command exits non-zero on failure.)

Then check the signature like before:

```
$ echo eyJhbGciOiJFUzI1NiIsImtpZCI6ImE3RW5ZbEI1QXp1S184ckNOZUlPZVJ2blRpU2VwM3Z6TXkwVDc1VzFBelEiLCJ4NWMiOlsiTUlJQ0ZUQ0NBYnVnQXdJQkFnSVJBS2Q2WUFpU0s1U2pZTEN2dnZxQnhvc3dDZ1lJS29aSXpqMEVBd0l3SHpFZE1Cc0dBMVVFQXhNVVZHVnpkQ0JKYm5SbGNtMWxaR2xoZEdVZ1EwRXdIaGNOTWpBd01UQTVNakF6T0RJMVdoY05NakF3TVRFd01qQXpPREkxV2pBV01SUXdFZ1lEVlFRREV3dGpiMlJsTFhOcFoyNWxjakJaTUJNR0J5cUdTTTQ5QWdFR0NDcUdTTTQ5QXdFSEEwSUFCRlN5SDFyekNVeHFCbXY0aVd2MTRmRTdQT0ROc2hsb0ljZ0l3Y2dpTk90MGhqcCtGNWl0QkJWaTlPM1Y1SXc5MFJWWG9VazhERmdjOVJPUjQybEpHN1NqZ2VBd2dkMHdEZ1lEVlIwUEFRSC9CQVFEQWdXZ01CMEdBMVVkSlFRV01CUUdDQ3NHQVFVRkJ3TUJCZ2dyQmdFRkJRY0RBakFkQmdOVkhRNEVGZ1FVQ0xrVFNoN2wyUE9HU2xjRzdrcVlvc1czMnhvd0h3WURWUjBqQkJnd0ZvQVVjaXVWbHNDSzJGYzV6K1hTcC91UCtKVjMzdlV3RmdZRFZSMFJCQTh3RFlJTFkyOWtaUzF6YVdkdVpYSXdWQVlNS3dZQkJBR0NwR1RHS0VBQkJFUXdRZ0lCQVFRUWJXbHJaVUJsZUdGdGNHeGxMbU52YlFRck5sQlVRMjFpYUc5Q01tVXdabVpWZWpCdFVVSm1hek51ZEdvelUxUmpSbmg2UkZwNVYzSTRlVWhoV1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThHYUdNcDJBbFJTcFB3MEQ0ZGVpaGw4RjJlczlHMHZIMVFHQnl6bG5mUUFDSUhYcXoyL1c2TTJkd1dUdC9TUzc3SElueEU3WklWdmhLa3hDcW1kRUtIdS8iLCJNSUlCbWpDQ0FVQ2dBd0lCQWdJUWZTMVlFNVRvMUFxQkNFYzN5U2RYTURBS0JnZ3Foa2pPUFFRREFqQVhNUlV3RXdZRFZRUURFd3hVWlhOMElGSnZiM1FnUTBFd0hoY05NakF3TVRBNU1UZzFPVEUxV2hjTk16QXdNVEEyTVRnMU9URTFXakFmTVIwd0d3WURWUVFERXhSVVpYTjBJRWx1ZEdWeWJXVmthV0YwWlNCRFFUQlpNQk1HQnlxR1NNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQUJKRDdta25GSTJpQkJRT3RzdmtDdGs1Qm4xUitOWkdEaG11WENXUWs5RHJyYjRzMXJjZ3BPZy9OSHV2VlpUcGZMeGpzWEdtVk9RYWJMWnFHWG9RbDZwYWpaakJrTUE0R0ExVWREd0VCL3dRRUF3SUJCakFTQmdOVkhSTUJBZjhFQ0RBR0FRSC9BZ0VBTUIwR0ExVWREZ1FXQkJSeUs1V1d3SXJZVnpuUDVkS24rNC80bFhmZTlUQWZCZ05WSFNNRUdEQVdnQlNZdlozbVRHZDF5T2Zrc09QUzV4U2VBdEtkZ1RBS0JnZ3Foa2pPUFFRREFnTklBREJGQWlFQThJU25reVo4QXVQT3Q1NTNVZnpMSDBBS2ZuYytqTDhxN0ozdUVYbkRCUmNDSUdWVWN2OE1uV0VXS0tPNUpaQitqVDlpV3d5KzJPZm5tbHRrNU45VHFRQnUiXX0.OTM2NGNkMmI2OTI5Njk3MmIwZWVhYmUyNGY5NTE2MzM2Njc2ZGE0NTY4OTUwNzY1NWMyZjFkYjllYTY5ZDFiNCAgc3RlcAo.OGmOogt9WK17S474gAlBZg0-Jog7eKG4fjYEqAErTaRj7CA6odhlvmEqhveFfk1EgsWzxbMe2kqAw_4oD_GITg \
  | step crypto jws verify --key <(step certificate key signer.crt) \
  | awk '{print $1}' \
  | xargs -Istep-hash step crypto hash compare step-hash step
ok
```

**Comment by CMCDragonkai** [comment]
In this case, I have to use x509 certificates because that's what the client side has to verify artifacts. So while there other signature tech, we need to use the x509 standard.

**Comment by CMCDragonkai** [comment]
For now if you guys add in the code signing certs. That would be enough for us to use the openssl cms commands for signing and verification.

**Comment by dopey** [comment]
Hey @CMCDragonkai, sorry for the delay on the "ETA for next release" question. Was out for the holidays :) 

Realistically this won't appear in master for probably another month. A release would follow shortly after that. This is because our aim would be to push this out as a larger change that would allow greater configurability for many more features of x509 certificates.

However, the change to enable the code signing extension for all new certificates is actually really small. If you need this sooner, I'd be happy to help you get setup with a working branch.

**Comment by CMCDragonkai** [comment]
Regarding #200, I'd really hope you consider standard x509 openssl cms commands before going for JWT code signing. The openssl stuff is more "standardized".

Yea didn't someone provide instructions above about how to add in x509 code signing feature?

**Comment by dopey** [comment]
Yep, that was me. Just wanted to follow up on the question of ETA.

**Comment by Prathima-reddy** [comment]
Gvb

**Comment by dopey** [comment]
I believe this is not possible with the recent release of certificate flexibility templates. If you feel I'm closing this issue prematurely, let me know and I'll happily reopen.
